input-output-hk/lace
# feat(extension): LW-6779 on-demand addresses discovery

# Checklist

- [x] JIRA - [LW-6779](https://input-output.atlassian.net/browse/LW-6779), [W-8841](https://input-output.atlassian.net/browse/LW-8841), [LW-8842](https://input-output.atlassian.net/browse/LW-8842), [LW-8843](https://input-output.atlassian.net/browse/LW-8843)

---

## Proposed solution

**A button in settings triggering a address discovery**

At the current point in time, there is no simple way to use the SDK API to trigger the HD wallet discovery. Therefore a hacky way has been developed that achieves the following goals:
 - allow for triggering the discovery on demand
 - catch the discovery status initiated by the PersonalWallet or on-demand request
 - catching any errors happening during the discovery

Implementation details:
There is an `addressesDiscoverer` logic living in the service worker. It consists of:
 - the `addressesDiscoverer` public interface accessible via remote API for the frontend instances. It exposes
   - `setup` method for configuring the `addressesDiscoverer` for specific network
   - `discover` method for triggering the discovery on-demand
   - `status$` stream providing the current discovery status
 - the `getDiscovererInstance` remembering and returning an instance of a discovery logic (wrapped `HDSequentialDiscovery` object) for a given network. That instance gets accessed during the creation of the `PersonalWallet` instance and during the on-demand discovery request using the `addressesDiscoverer` API.

Every time the network changes the `addressesDiscoverer` gets updated with the new `chainName` and the new asyncKeyAgent which is being exposed to the `addressesDiscoverer` through the remote API.

## Testing

Scenario 1:
1. Install extension on two browsers/accounts
2. restore the same wallet in both places
4. make a delegation in the first instance to at least one more pool and wait till it finishes
6. go to staking in the second instance and confirm there is one pool missing
7. go to settings
8. click sync.
9. confirm you have all the pools

Scenario 2:
1. Restore a wallet and click go to my wallet button fast enough
2. see the sync overlay

## Screenshots

Settings option:
![image](https://github.com/input-output-hk/lace/assets/9270461/7574f15a-09b8-47be-93f3-befed50e7de6)
<img width="981" alt="image" src="https://github.com/input-output-hk/lace/assets/9270461/4dd6c0c8-b713-48a9-832f-5cfde5792cf4">
Overlay:
![image](https://github.com/input-output-hk/lace/assets/9270461/2a9afa94-59df-4cae-a2b3-31c28e63c46e)
<img width="981" alt="image" src="https://github.com/input-output-hk/lace/assets/9270461/5c97ff93-a68c-43d4-83fe-8248d1284dbe">
Success toast:
![image](https://github.com/input-output-hk/lace/assets/9270461/965b70ad-9731-4621-8ec7-054f23eb31ad)
<img width="981" alt="image" src="https://github.com/input-output-hk/lace/assets/9270461/4f647093-5ab8-4c38-a2b4-38523c035645">
Error toast:
![image](https://github.com/input-output-hk/lace/assets/9270461/6db7928c-b921-43f4-bea3-a072ba5c79c8)
<img width="937" alt="image" src="https://github.com/input-output-hk/lace/assets/9270461/30fc7536-64c3-43f8-8793-29595d0763c8">


---
# Allure report
`allure-report-publisher` generated test report!



**smokeTests**: ❌ [test report](https://lace-qa:8PP5wBaDV6UoXj@dq4ajm5i5q7bz.cloudfront.net/linux/chrome/2725/6660158911/index.html) for [e2a0291d](https://github.com/input-output-hk/lace/pull/673/commits/e2a0291d2341a04da03e64fe94342f35981140b5)
|       | passed | failed | skipped | flaky | total | result |
|-------|--------|--------|---------|-------|-------|--------|
| Total | 3      | 29     | 0       | 0     | 32    | ❌     |





[LW-6779]: https://input-output.atlassian.net/browse/LW-6779?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[LW-8842]: https://input-output.atlassian.net/browse/LW-8842?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[LW-8843]: https://input-output.atlassian.net/browse/LW-8843?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
